lib/ASTGen/Sources/ASTGen/Bridge.swift

```diff
@@ -34,6 +34,7 @@ extension BridgedSourceLoc {
   ) {
     if let start = buffer.baseAddress,
       position.utf8Offset >= 0 && position.utf8Offset < buffer.count {
+      assert(position.utf8Offset >= 0 && position.utf8Offset < buffer.count)
```

Comment:
This assertion will never be tripped.

Reply:
is this `if-else` statement ever even reachable, because the issue says neither is ever expected? So how would we assert for something in a place that it is never tripped.

Reply:
There's a difference between expectation and reality though, right? We want to assert precisely because we never expect this condition to fail, instead of retuning a null location whenever it happens to fail. In other words, we want this condition to become part of the initializer's contract.

Currently, the assertion is placed along a path where the asserted condition is known to be true, and will never fulfill its purpose.

Reply:
so the `init` of `BridgedSourceLoc` becomes

```swift
  init(
    at position: AbsolutePosition,
    in buffer: UnsafeBufferPointer<UInt8>,
    ast: Bool
  ) {
```
And whenever we are declaring any `BridgedSourceLoc` we would need to pass in the `assert`? Which in this case is checking the position bounds?

Reply:
Not quite. What I mean is a plain old assertion inside the `init`. Currently, we construct a null location when the position is out of bounds. Instead, we want to assert that the given position is within bounds right from the start, and not construct a null location at all.

Did the word "contract" mislead you? When we place expectations on a function’s inputs using assertions due to not being able to enforce them at compile time, they become part of the function’s run-time contract. The contract is simply a user agreement, something that one accepts by calling the function, and that may or may not be fully documented. For instance, suppose a function is documented to do something, and to require an integer argument > 1. The developer assumes responsibility for the documented behavior by vending the function, and the user assumes responsibility for satisfying the > 1 requirement by using the function.

Reply:
The user, based on the contract, is "agreeing" to pass a position that is already in bounds. The `init` right now is checking for validity of the position with the `if-else`, but we are "expecting" that it will always be set to a valid location and NEVER reach the `else`, which sets it the location to `null`. 

But, in reality we are never 100% that the position will ALWAYS be in bounds, so we are asserting from the get go of `init` and that way it becomes a part of the function's run time contract and errors out?

Reply:
Yep!

@ahoppen Did you suggest an assert rather than a precondition because this is a hot path?

Reply:
I said assertion because I didn’t actually think about whether I preferred an assertion or precondition. I just chatted to @bnbarham  and @rintaro and we prefer `precondition` and want to follow the same paradigm as in swift-syntax

https://github.com/apple/swift-syntax/blob/806c4732347d34b645ec8ec357914d6c3e3716ba/Sources/SwiftSyntax/Assert.swift#L13-L23

Reply:
The `precondition` will be checking the bounds of the buffer and the location will be set to a `non-null` value if it passes.

```swift
extension BridgedSourceLoc {
  /// Form a source location at the given absolute position in `buffer`.
  init(
    at position: AbsolutePosition,
    in buffer: UnsafeBufferPointer<UInt8>
  ) {
    precondition(position.utf8Offset >= 0 && position.utf8Offset < buffer.count)
    self = SourceLoc_advanced(BridgedSourceLoc(raw: buffer.baseAddress), SwiftInt(position.utf8Offset))
  }
}
```

Reply:
That looks good. Can you push your changes?